src/codecatalyst/devEnv.ts

```diff
@@ -0,0 +1,206 @@
+            shutdownMessage.setupMessage(maxInactivityMinutes, userIsActive, () =>
+                warningMessage.clearExistingMessage()
+            )
+        })
+
+        devEnv.updateUserActivity()
+    }
+}
+
+export class InactivityWarningMessage {
+    private beforeWarningMessageShownTimeout: NodeJS.Timeout | undefined
+    private currentWarningMessageTimeout: Timeout | undefined
+
+    static #instance: InactivityWarningMessage | undefined
+
+    static get instance(): InactivityWarningMessage {
+        return (this.#instance ??= new InactivityWarningMessage())
+    }
+
+    /**
+     * @WARNING Only use this for testing purposes, use {@link instance} otherwise.
+     *
+     * @param relativeMinuteMillis What a minute is in the context of testing
+     */
+    constructor(private readonly relativeMinuteMillis?: number) {}
+
+    async setupMessage(maxInactivityMinutes: number, userIsActive: () => Promise<any> | any): Promise<void> {
+        this.clearExistingMessage()
+
+        const showMessageAfterMinutes = this.getMinutesUntilShowWarningMessage(maxInactivityMinutes)
+
+        this.beforeWarningMessageShownTimeout = globals.clock.setTimeout(() => {
+            const inactiveMinutes = showMessageAfterMinutes
+            const remainingMinutes = maxInactivityMinutes - inactiveMinutes
+            this.showInactivityWarningMessage(inactiveMinutes, remainingMinutes, userIsActive)
+        }, showMessageAfterMinutes * (this.relativeMinuteMillis ?? 60 * 1000))
+    }
+
+    clearExistingMessage() {
+        if (this.beforeWarningMessageShownTimeout) {
+            clearTimeout(this.beforeWarningMessageShownTimeout)
+            this.beforeWarningMessageShownTimeout = undefined
+        }
+
+        if (this.currentWarningMessageTimeout) {
+            this.currentWarningMessageTimeout.dispose()
+            this.currentWarningMessageTimeout = undefined
+        }
+    }
+
+    private getMinutesUntilShowWarningMessage(maxInactivityMinutes: number): number {
+        if (maxInactivityMinutes >= 10) {
+            // Show warning messages 5 minutes before shutdown
+            return maxInactivityMinutes - 5
+        } else {
+            // Show warning message 3 minutes before shutdown
+            return maxInactivityMinutes - 3
+        }
+    }
+
+    /**
+     * Shows a cancellable message warning about the remaining inactive minutes till shutdown.
+     *
+     * It is updated each {@link interval} with the latest remaining minutes.
+     */
+    private showInactivityWarningMessage(
+        inactiveMinutes: number,
+        remainingMinutes: number,
+        userIsActive: () => Promise<any> | any,
+        interval: Timeout = this.warningMessageInterval()
+    ) {
+        if (remainingMinutes <= 1) {
+            // Recursive base case, with only 1 minute left we do not want to show warning messages anymore.
+            interval.dispose()
+            return
+        }
```

Comment:
And that's because we show the final shutdown message right? That's a sign that the two classes shouldn't be separate.

Reply:
Good call, refactored so the message is in this call